### ISSUE TITLE
Singleton nomemclature not stable across analyses

### ISSUE THREAD

**Issue by marchoeppner** [insapathogenomics/ReporTree#35]
Hi,

so I have been trying to replicate a typical incremental analysis of cgMLST allele profiles to develop a database that reflects the data structure of ReporTree. In that process, I noticed that singleton nomenclature is not stable across analyses (latest version of ReportTree, 2.5.3). 

What I have done:

Take 150 cgMLST profiles, in batches of 50 (from https://github.com/ncezid-biome/datasets/tree/master/datasets).

Batch 1 (first 50 profiles) was analysed without a nomemclature file, to create a new partitions files. 

Batch 2 was analyzed with 100 profiles (from batch 1, plus 50 more) with the partitions file from batch 1  to carry over nomenclature. 

Batch 3 was analyzed with 150 profiles (all from batch 1 und 2, plus 50 more) with the partitions file from batch 2. 

The cluster nomemclature appears stable:

```
Batch 3: grep "cluster_1     " output/Se_clusterComposition.tsv 
MST-10x1.0      cluster_1       7       FDA00013489,FDA00013488,FDA00013487,FDA00013494,FDA00013492,FDA00013491,FDA00013490
MST-20x1.0      cluster_1       9       FDA00013489,FDA00013488,FDA00013487,FDA00013494,PNUSAS044147,FDA00013493,FDA00013492,FDA00013491,FDA00013490
Batch 2: grep "cluster_1     " ../2/output/Se_clusterComposition.tsv 
MST-10x1.0      cluster_1       7       FDA00013492,FDA00013489,FDA00013491,FDA00013488,FDA00013494,FDA00013487,FDA00013490
MST-20x1.0      cluster_1       8       FDA00013493,FDA00013492,FDA00013489,FDA00013491,FDA00013488,FDA00013494,FDA00013487,FDA00013490
Batch 1: grep "cluster_1     " ../1/output/Se_clusterComposition.tsv 
MST-10x1.0      cluster_1       7       FDA00013491,FDA00013487,FDA00013492,FDA00013490,FDA00013488,FDA00013489,FDA00013494
MST-20x1.0      cluster_1       8       FDA00013491,FDA00013487,FDA00013492,FDA00013490,FDA00013493,FDA00013488,FDA00013489,FDA00013494
```

The singletons are not:

```
Batch 3: grep "singleton_1   " output/Se_clusterComposition.tsv
MST-10x1.0      singleton_1     1       PNUSAS040688
MST-20x1.0      singleton_1     1       PNUSAS038526
Batch 2: grep "singleton_1   " ../2/output/Se_clusterComposition.tsv 
MST-10x1.0      singleton_1     1       FDA00013493
MST-20x1.0      singleton_1     1       PNUSAS038526
Batch 1: grep "singleton_1   " ../1/output/Se_clusterComposition.tsv 
MST-10x1.0      singleton_1     1       FSIS11807011
MST-20x1.0      singleton_1     1       FSIS31800132

```

Is this known, or am I possibly doing something wrong? Happy to share the data for this.

**Comment by marchoeppner** [comment]
Ah you know what, here is the data, just in case. I am using Singularity, so you'll have to adjust the command line calls in run.sh, obviously. 
[reportree_singleton_not_stable.zip](https://github.com/user-attachments/files/17988202/reportree_singleton_not_stable.zip)

**Comment by vmixao** [comment]
Hi @marchoeppner,

Thank you for your interest in ReporTree. Thanks also for sharing the data, which was useful for us to check what was going on.

ReporTree is keeping the nomenclature between runs for clusters and singletons, as expected and as you can see, for example, in the _metadata_w_partitions.tsv_ or the _nomenclature_changes.tsv_ files. What you noticed is a particular issue that affected the _clusterComposition.tsv_ file (and only this file), which was not being updated with the nomenclature provided by the user. Thanks for pointing this out. We have now solved this (commit 90b2b21) in ReporTree v2.5.4.

We take this opportunity to bring to your attention some issues encountered while checking your outputs. Specifically:

- We realized that your allele matrix still contains the different codes of chewBBACA for non-called alleles (e.g. LNF). These codes should be replaced by 0 in order to be considered as missing data. Otherwise, ReporTree and any other clustering tool will count them as an allele. Of note, ReporTree is able to deal with the INF-* codes that you may find in your matrix, if populating the schema. You just need to worry about the codes that correspond to the absence of an allele called.
- Once you have your cleaned matrix with missing data, it is important to perform the analysis with a cgMLST schema. For this, you can either rely on the ‘--site-inclusion’ argument of ReporTree, which will be used by ReporTree to determine the cgMLST according to the value you specified (with this strategy, the cgMLST is dynamic and specific of a dataset). Or, alternatively, you can indicate a static list of loci to be used in the analysis (e.g. the list of cgMLST loci of the Salmonella INNUENDO schema), which is a better approach for nomenclature maintenance and continuous surveillance scenarios. Of note, this last approach can be combined with a zoom-in analysis in clusters of interest (e.g. outbreak-level), in which the cgMLST schema will be extended with the accessory loci specific of that cluster, thus providing a higher resolution.

Hope this helps you implement ReporTree in your daily work.

Kind regards,
